core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,1164 @@
+        }
+    }
+
+    fn lockdown(&self) {
+        for state in self.state.iter() {
+            let mut lockguard = state.lock().unwrap();
+            // block reception of votes
+            lockguard.reset();
+        }
+    }
+
+    /// Sends mock packets to everyone in the cluster
+    fn sender_thread(
+        state: Arc<[Mutex<SharedState>; 3]>,
+        packet_size: Arc<AtomicU16>,
+        cluster_info: Arc<ClusterInfo>,
+        mut epoch_specs: EpochSpecs,
+        socket: Arc<UdpSocket>,
+        command: Receiver<Command>,
+    ) {
+        let mut packet_buf = vec![0u8; MOCK_VOTE_PACKET_MAX_SIZE];
+        let id = cluster_info.id();
+        for command in command.iter() {
+            let (slot, votor_msg) = match command {
+                Command::SendNotarize(slot) => (slot, VotorMessageType::Notarize),
+                Command::SendNotarizeCertificate(slot) => {
+                    (slot, VotorMessageType::NotarizeCertificate)
+                }
+                Command::SendFinalize(slot) => (slot, VotorMessageType::Finalize),
+                Command::SendFinalizeCertificate(slot) => {
+                    (slot, VotorMessageType::FinalizeCertificate)
+                }
+            };
+            let packet_size = (packet_size.load(Ordering::Relaxed) as usize)
+                .clamp(MOCK_VOTE_HEADER_SIZE, MOCK_VOTE_PACKET_MAX_SIZE);
+
+            prep_and_sign_packet(
+                &mut packet_buf,
+                slot,
+                votor_msg,
+                cluster_info.keypair().as_ref(),
+            );
+
+            // prepare addresses to send the packets
+            let mut send_instructions = Vec::with_capacity(2000);
+            {
+                let mut state = get_state_for_slot(state.as_slice(), slot).lock().unwrap();
+                // check if our task was aborted, avoid sending if it was.
+                if state.current_slot != slot {
+                    return;
+                }
+
+                for (peer, info) in state.peers.iter() {
+                    send_instructions.push((&packet_buf[0..packet_size], info.address));
+                    trace!(
+                        "{id}: send {votor_msg:?} for slot {slot} to {} for {peer}",
+                        info.address
+                    );
+                }
+            }
+            // broadcast to everybody at once
+            batch_send(&socket, send_instructions);
+        }
+    }
+
+    pub(crate) fn signal_new_slot(&mut self, slot: Slot, root_bank: &Bank) {
+        let Some(config) = get_test_config_from_account::<TestConfig>(root_bank) else {
+            return; // no config is available => test can not run
+        };
+        // copy basic parameters from the config
+        self.packet_size
+            .store(config.packet_size, Ordering::Relaxed);
+        self.verify_signatures
+            .store(config.verify_signatures, Ordering::Relaxed);
+        let interval = config.test_interval_slots as u64;
```

Comment:
Sorry about going back to this again. Do you think it would help to add config.test_participate_ratio as well? So let's say if 2000 node tests somehow failed, we can reduce it to 1000 node tests to see whether this helps. For example, config.test_participate_ratio = 0.5, then there is only half chance that I send out packets at any time, I still receive packets from everywhere, but I would calculate the threshold * config.test_partipate_ratio (probably a bit lower actually) to determine the new stake threshold.

So we can either start every 1000 slots -> every 100 slots -> every 10 slots -> every slot, or
ratio = 0.001 -> 0.01 -> 0.1 -> 0.5 -> 1
Test different patterns.

What do you think?

Reply:
Ok I will put it together in a follow-up PR on top of this, since this one is complicated enough.